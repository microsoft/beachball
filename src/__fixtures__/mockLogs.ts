import { jest, afterEach, beforeAll, afterAll } from '@jest/globals';
import type { SpyInstance } from 'jest-mock';

/** Methods that will be mocked. More could be added later if needed. */
type MockLogMethod = 'log' | 'warn' | 'error';
const mockedMethods: MockLogMethod[] = ['log', 'warn', 'error'];

type MockLogsOptions = {
  /**
   * Whether to also log to the real console (or subset of methods to log to the console).
   * All logging can be enabled by setting the VERBOSE env var.
   */
  alsoLog?: boolean | MockLogMethod[];
};

export type MockLogs = {
  /** Mocked methods (to access calls etc) */
  mocks: { [k in MockLogMethod]: SpyInstance<typeof console.log> };

  /** Set override options for one test only */
  setOverrideOptions: (options: MockLogsOptions) => void;

  /** Call `mockClear` on all mocks to reset calls */
  clear: () => void;

  /** Get the lines logged to a particular method (or all methods) */
  getMockLines: (method: MockLogMethod | 'all', sanitizeGuids?: boolean) => string;
};

/**
 * Initialize console log mocks, which will be reset after each test. This should be called **outside**
 * of any lifecycle hooks or tests because it calls lifecycle hooks internally for setup and teardown.
 */
export function initMockLogs(options: MockLogsOptions = {}): MockLogs {
  const { alsoLog } = options;
  let allLines: unknown[][] = [];
  let overrideOptions: MockLogsOptions | undefined;
  const jestConsole = { ...console };

  const logs: MockLogs = {
    mocks: {} as MockLogs['mocks'],
    setOverrideOptions: options => {
      overrideOptions = options;
    },
    clear: () => {
      allLines = [];
      overrideOptions = undefined;
      Object.values(logs.mocks).forEach(mock => mock.mockClear());
    },
    getMockLines: (method, sanitizeGuids) => {
      const lines = (method === 'all' ? allLines : logs.mocks[method].mock.calls)
        .map(args =>
          args
            .join(' ')
            .split('\n')
            .map(l => l.trimEnd()) // prevent trailing whitespace in snapshots
            .join('\n')
        )
        .join('\n')
        .trim();

      return sanitizeGuids
        ? // replace change file name GUIDs
          lines.replace(/[a-f\d]{8}(-[a-f\d]{4}){3}-[a-f\d]{12}/g, '<guid>')
        : lines;
    },
  };

  beforeAll(() => {
    for (const method of mockedMethods) {
      const mainShouldLog = shouldLog(method, alsoLog);

      logs.mocks[method] = jest.spyOn(console, method).mockImplementation((...args) => {
        const currentShouldLog =
          overrideOptions === undefined ? mainShouldLog : shouldLog(method, overrideOptions.alsoLog);
        allLines.push([`[${method}]`, ...args]);
        if (process.env.VERBOSE || currentShouldLog) {
          jestConsole[method](...args);
        }
      });
    }
  });

  afterEach(() => {
    logs.clear();
  });

  afterAll(() => {
    Object.values(logs.mocks).forEach(mock => mock.mockRestore());
  });

  return logs;
}

function shouldLog(method: MockLogMethod, alsoLog: boolean | MockLogMethod[] | undefined) {
  return typeof alsoLog === 'boolean' ? alsoLog : alsoLog?.includes(method);
}
