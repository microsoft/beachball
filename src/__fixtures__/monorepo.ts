import path from 'path';
import * as fs from 'fs-extra';
import { BeachballOptions } from '../types/BeachballOptions';
import { BaseRepositoryFactory, Repository } from './repository';
import { PackageJson } from '../types/PackageInfo';

export const packageJsonFixtures: { [path: string]: PackageJson } = {
  'packages/foo': {
    name: 'foo',
    version: '1.0.0',
    dependencies: {
      bar: '^1.3.4',
    },
    main: 'src/index.ts',
    onPublish: {
      main: 'lib/index.js',
    },
    afterPublish: {
      notify: 'message',
    },
  } as PackageJson,

  'packages/bar': {
    name: 'bar',
    version: '1.3.4',
    dependencies: {
      baz: '^1.3.4',
    },
  },

  'packages/baz': {
    name: 'baz',
    version: '1.3.4',
  },

  'packages/grouped/a': {
    name: 'a',
    version: '3.1.2',
  },

  'packages/grouped/b': {
    name: 'b',
    version: '3.1.2',
    dependencies: ['bar'] as any,
  },
};

const beachballConfigFixture = {
  groups: [
    {
      disallowedChangeTypes: null,
      name: 'grouped',
      include: 'grouped*',
    },
  ],
} as BeachballOptions;

export class MonoRepoFactory extends BaseRepositoryFactory {
  constructor() {
    super('beachball-monorepo-upstream-');
  }

  protected initFixture(tmpRepo: Repository) {
    for (const pkg of Object.keys(packageJsonFixtures)) {
      const packageJsonFixture = packageJsonFixtures[pkg];
      const packageJsonFile = path.join(pkg, 'package.json');

      fs.mkdirpSync(path.join(tmpRepo.rootPath, pkg));

      fs.writeJSONSync(path.join(tmpRepo.rootPath, packageJsonFile), packageJsonFixture, {
        spaces: 2,
      });
      tmpRepo.commitChange(packageJsonFile);
    }

    tmpRepo.commitChange('package.json', JSON.stringify({ name: 'monorepo-fixture', version: '1.0.0' }, null, 2));

    tmpRepo.commitChange('beachball.config.js', 'module.exports = ' + JSON.stringify(beachballConfigFixture, null, 2));
  }
}
