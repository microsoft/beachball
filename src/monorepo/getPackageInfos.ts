import { findPackageRoot, findGitRoot } from '../paths';
import fs from 'fs-extra';
import path from 'path';
import { listAllTrackedFiles } from '../git';
import { PackageInfos } from '../types/PackageInfo';
import { infoFromPackageJson } from './infoFromPackageJson';

export function getPackageInfos(cwd: string) {
  const gitRoot = findGitRoot(cwd)!;
  const packageJsonFiles = listAllTrackedFiles(['**/package.json', 'package.json'], gitRoot);
  const packageInfos: PackageInfos = {};

  if (packageJsonFiles && packageJsonFiles.length > 0) {
    packageJsonFiles.forEach(packageJsonPath => {
      try {
        const packageJsonFullPath = path.join(gitRoot, packageJsonPath);
        const packageJson = fs.readJSONSync(packageJsonFullPath);
        packageInfos[packageJson.name] = infoFromPackageJson(packageJson, packageJsonFullPath);
      } catch (e) {
        // Pass, the package.json is invalid
        console.warn(`Invalid package.json file detected ${packageJsonPath}: `, e);
      }
    });
  } else {
    const packageJsonFullPath = path.join(gitRoot, findPackageRoot(cwd)!, 'package.json');
    const packageJson = fs.readJSONSync(packageJsonFullPath);
    packageInfos[packageJson.name] = infoFromPackageJson(packageJson, packageJsonFullPath);
  }
  return packageInfos;
}
