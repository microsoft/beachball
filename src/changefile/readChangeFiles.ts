import { ChangeSet, ChangeInfo } from '../types/ChangeInfo';
import { getChangePath } from '../paths';
import fs from 'fs-extra';
import path from 'path';
import { BeachballOptions } from '../types/BeachballOptions';
import { getScopedPackages } from '../monorepo/getScopedPackages';
import { getFileAddedHash, getChangesBetweenRefs } from '../git';

export function readChangeFiles(options: BeachballOptions): ChangeSet {
  const { path: cwd } = options;
  const scopedPackages = getScopedPackages(options);
  const changeSet: ChangeSet = new Map();
  const changePath = getChangePath(cwd);
  const fromRef = options.fromRef;

  if (!changePath || !fs.existsSync(changePath)) {
    return changeSet;
  }

  const allChangeFiles = fs.readdirSync(changePath);
  const filteredChangeFiles: string[] = [];

  if (fromRef) {
    const changeFilesSinceFromRef = getChangesBetweenRefs(
      fromRef,
      'HEAD',
      [
        '--diff-filter=d', // excluding deleted files from the diff.
        '--relative', // results will include path relative to the cwd, i.e. only file names.
      ],
      changePath
    );

    allChangeFiles
      .filter(fileName => changeFilesSinceFromRef?.includes(fileName))
      .forEach(fileName => filteredChangeFiles.push(fileName));
  } else {
    filteredChangeFiles.push(...allChangeFiles);
  }

  try {
    // sort the change files by modified time. Most recent modified file comes first.
    filteredChangeFiles.sort(function(f1, f2) {
      return (
        fs.statSync(path.join(changePath, f2)).mtime.getTime() - fs.statSync(path.join(changePath, f1)).mtime.getTime()
      );
    });
  } catch (err) {
    console.warn('Failed to sort change files', err);
  }

  filteredChangeFiles.forEach(changeFile => {
    try {
      const changeFilePath = path.join(changePath, changeFile);
      const changeInfo: ChangeInfo = {
        ...fs.readJSONSync(changeFilePath),
        // Add the commit hash where the file was actually first introduced
        commit: getFileAddedHash(changeFilePath, cwd) || '',
      };

      const packageName = changeInfo.packageName;
      if (scopedPackages.includes(packageName)) {
        changeSet.set(changeFile, changeInfo);
      } else {
        console.log(`Skipping reading change file for out-of-scope package ${packageName}`);
      }
    } catch (e) {
      console.warn(`Invalid change file detected: ${changeFile}`);
    }
  });
  return changeSet;
}
