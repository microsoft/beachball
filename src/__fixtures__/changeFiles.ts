import fs from 'fs';
import path from 'path';
import { writeChangeFiles } from '../changefile/writeChangeFiles';
import { getChangePath } from '../paths';
import { ChangeFileInfo } from '../types/ChangeInfo';

/** Change file with `packageName` required and other props optional */
export type PartialChangeFile = { packageName: string } & Partial<ChangeFileInfo>;

/**
 * Generates and writes change files for the given packages.
 * @param changes Array of package names or partial change files (which must include `packageName`).
 * Default values are `type: 'minor'`, `dependentChangeType: 'patch'`, and placeholders for other fields.
 * @param cwd Working directory
 * @param groupChanges Whether to group all changes into one change file.
 */
export function generateChangeFiles(changes: (string | PartialChangeFile)[], cwd: string, groupChanges?: boolean) {
  writeChangeFiles({
    changes: changes.map(change => {
      change = typeof change === 'string' ? { packageName: change } : change;
      return {
        comment: `${change.packageName} test comment`,
        email: 'test@test.com',
        type: 'minor',
        dependentChangeType: 'patch',
        ...change,
      };
    }),
    groupChanges,
    cwd,
  });
}

/** Get full paths to existing change files under `cwd` */
export function getChangeFiles(cwd: string): string[] {
  const changePath = getChangePath(cwd);
  return changePath && fs.existsSync(changePath) ? fs.readdirSync(changePath).map(p => path.join(changePath, p)) : [];
}
