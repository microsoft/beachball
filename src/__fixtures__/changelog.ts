import fs from 'fs-extra';
import path from 'path';
import _ from 'lodash';
import { SortedChangeTypes } from '../changefile/changeTypes';
import { ChangelogJson } from '../types/ChangeLog';

/** Read the CHANGELOG.md under the given package path, sanitizing any dates for snapshots */
export function readChangelogMd(packagePath: string): string {
  const changelogFile = path.join(packagePath, 'CHANGELOG.md');
  const text = fs.readFileSync(changelogFile, { encoding: 'utf-8' });
  return text.replace(/\w\w\w, \d\d \w\w\w [\d :]+?GMT/gm, '(date)');
}

/** Read the CHANGELOG.json under the given package path */
export function readChangelogJson(packagePath: string, cleanForSnapshot: boolean = false): ChangelogJson {
  const changelogJsonFile = path.join(packagePath, 'CHANGELOG.json');
  const json = fs.readJSONSync(changelogJsonFile, { encoding: 'utf-8' });
  return cleanForSnapshot ? cleanChangelogJson(json) : json;
}

/**
 * Clean changelog json for a snapshot: replace dates and SHAs with placeholders.
 * Note: this clones the changelog object rather than modifying the original.
 */
export function cleanChangelogJson(changelog: ChangelogJson): ChangelogJson {
  changelog = _.cloneDeep(changelog);
  // for a better snapshot, make the fake commit match if the real commit did
  const fakeCommits: { [commit: string]: string } = {};
  let fakeHashNum = 0;

  for (const entry of changelog.entries) {
    entry.date = '(date)';
    for (const changeType of SortedChangeTypes) {
      if (entry.comments[changeType]) {
        for (const comment of entry.comments[changeType]!) {
          if (!fakeCommits[comment.commit]) {
            fakeCommits[comment.commit] = `(sha1-${fakeHashNum++})`;
          }
          comment.commit = fakeCommits[comment.commit];
        }
      }
    }
  }
  return changelog;
}
