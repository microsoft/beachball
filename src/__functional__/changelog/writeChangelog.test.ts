import { describe, expect, it, beforeAll, afterAll } from '@jest/globals';
import { generateChangeFiles } from '../../__fixtures__/changeFiles';
import { cleanChangelogJson, readChangelogJson, readChangelogMd } from '../../__fixtures__/changelog';
import { initMockLogs } from '../../__fixtures__/mockLogs';
import { RepositoryFactory } from '../../__fixtures__/repositoryFactory';

import { writeChangelog } from '../../changelog/writeChangelog';
import { getPackageInfos } from '../../monorepo/getPackageInfos';
import { readChangeFiles } from '../../changefile/readChangeFiles';
import { BeachballOptions } from '../../types/BeachballOptions';
import { ChangeFileInfo, ChangeInfo } from '../../types/ChangeInfo';

function getChange(packageName: string, comment: string): ChangeFileInfo {
  return {
    comment,
    email: 'test@testtestme.com',
    packageName,
    type: 'patch',
    dependentChangeType: 'patch',
  };
}

describe('writeChangelog', () => {
  let repositoryFactory: RepositoryFactory;
  let monoRepoFactory: RepositoryFactory;

  initMockLogs();

  beforeAll(() => {
    // These tests can share the same repo factories because they don't push to origin
    // (the actual tests run against a clone)
    repositoryFactory = new RepositoryFactory('single');
    monoRepoFactory = new RepositoryFactory('monorepo');
  });

  afterAll(() => {
    repositoryFactory.cleanUp();
    monoRepoFactory.cleanUp();
  });

  it('generates correct changelog', async () => {
    const repository = repositoryFactory.cloneRepository();
    repository.commitChange('foo');
    generateChangeFiles([getChange('foo', 'additional comment 2')], repository.rootPath);
    generateChangeFiles([getChange('foo', 'additional comment 1')], repository.rootPath);
    generateChangeFiles([getChange('foo', 'comment 1')], repository.rootPath);

    repository.commitChange('bar');
    generateChangeFiles([getChange('foo', 'comment 2')], repository.rootPath);

    const beachballOptions = { path: repository.rootPath } as BeachballOptions;
    const packageInfos = getPackageInfos(repository.rootPath);
    const changes = readChangeFiles(beachballOptions, packageInfos);

    await writeChangelog(beachballOptions, changes, { foo: 'patch' }, { foo: new Set(['foo']) }, packageInfos);

    expect(readChangelogMd(repository.rootPath)).toMatchSnapshot('changelog md');

    const changelogJson = readChangelogJson(repository.rootPath);
    expect(cleanChangelogJson(changelogJson)).toMatchSnapshot('changelog json');

    // Every entry should have a different commit hash
    const patchComments = changelogJson.entries[0].comments.patch!;
    const commits = patchComments.map(entry => entry.commit);
    expect(new Set(commits).size).toEqual(patchComments.length);

    // The first entry should be the newest
    expect(patchComments[0].commit).toBe(repository.getCurrentHash());
  });

  it('generates correct changelog in monorepo with groupChanges (grouped change FILES)', async () => {
    const monoRepo = monoRepoFactory.cloneRepository();
    monoRepo.commitChange('foo');
    const params = [monoRepo.rootPath, true /*groupChanges*/] as const;
    generateChangeFiles(
      [getChange('foo', 'additional comment 2'), getChange('bar', 'comment from bar change ')],
      ...params
    );
    generateChangeFiles([getChange('foo', 'additional comment 1')], ...params);
    generateChangeFiles([getChange('foo', 'comment 1')], ...params);

    monoRepo.commitChange('bar');
    generateChangeFiles([getChange('foo', 'comment 2')], ...params);

    const beachballOptions = { path: monoRepo.rootPath, groupChanges: true } as BeachballOptions;
    const packageInfos = getPackageInfos(monoRepo.rootPath);
    const changes = readChangeFiles(beachballOptions, packageInfos);

    await writeChangelog(beachballOptions, changes, { foo: 'patch', bar: 'patch' }, {}, packageInfos);

    // check changelogs for both foo and bar
    expect(readChangelogMd(monoRepo.pathTo('packages/foo'))).toMatchSnapshot('foo CHANGELOG.md');
    expect(readChangelogMd(monoRepo.pathTo('packages/bar'))).toMatchSnapshot('bar CHANGELOG.md');

    const fooJson = readChangelogJson(monoRepo.pathTo('packages/foo'));
    expect(cleanChangelogJson(fooJson)).toMatchSnapshot('foo CHANGELOG.json');
    expect(readChangelogJson(monoRepo.pathTo('packages/bar'), true /*clean*/)).toMatchSnapshot('bar CHANGELOG.json');

    // Every entry should have a different commit hash
    const patchComments = fooJson.entries[0].comments.patch!;
    const commits = patchComments.map(entry => entry.commit);
    expect(new Set(commits).size).toEqual(patchComments.length);

    // The first entry should be the newest
    expect(patchComments[0].commit).toBe(monoRepo.getCurrentHash());
  });

  it('generates correct grouped changelog', async () => {
    const monoRepo = monoRepoFactory.cloneRepository();
    monoRepo.commitChange('foo');
    generateChangeFiles([getChange('foo', 'comment 1')], monoRepo.rootPath);

    monoRepo.commitChange('bar');
    generateChangeFiles([getChange('bar', 'comment 2')], monoRepo.rootPath);
    generateChangeFiles([getChange('bar', 'comment 3')], monoRepo.rootPath);

    const beachballOptions: Partial<BeachballOptions> = {
      path: monoRepo.rootPath,
      changelog: {
        groups: [
          {
            masterPackageName: 'foo',
            changelogPath: monoRepo.rootPath,
            include: ['packages/foo', 'packages/bar'],
          },
        ],
      },
    };

    const packageInfos = getPackageInfos(monoRepo.rootPath);
    const changes = readChangeFiles(beachballOptions as BeachballOptions, packageInfos);

    await writeChangelog(beachballOptions as BeachballOptions, changes, {}, {}, packageInfos);

    // Validate changelog for foo and bar packages
    expect(readChangelogMd(monoRepo.pathTo('packages/foo'))).toMatchSnapshot('foo CHANGELOG.md');
    expect(readChangelogMd(monoRepo.pathTo('packages/bar'))).toMatchSnapshot('bar CHANGELOG.md');

    // Validate grouped changelog for foo and bar packages
    expect(readChangelogMd(monoRepo.rootPath)).toMatchSnapshot('grouped CHANGELOG.md');
  });

  it('generates grouped changelog without dependent change entries', async () => {
    const monoRepo = monoRepoFactory.cloneRepository();
    monoRepo.commitChange('baz');
    generateChangeFiles([getChange('baz', 'comment 1')], monoRepo.rootPath);

    const beachballOptions: Partial<BeachballOptions> = {
      path: monoRepo.rootPath,
      changelog: {
        groups: [
          {
            masterPackageName: 'foo',
            changelogPath: monoRepo.rootPath,
            include: ['packages/foo', 'packages/bar', 'packages/baz'],
          },
        ],
      },
    };

    const packageInfos = getPackageInfos(monoRepo.rootPath);
    const changes = readChangeFiles(beachballOptions as BeachballOptions, packageInfos);

    await writeChangelog(
      beachballOptions as BeachballOptions,
      changes,
      { bar: 'patch', baz: 'patch' },
      { bar: new Set(['baz']) },
      packageInfos
    );

    // Validate changelog for bar package
    const barChangelogText = readChangelogMd(monoRepo.pathTo('packages/bar'));
    expect(barChangelogText).toContain('- Bump baz');
    expect(barChangelogText).toMatchSnapshot('bar CHANGELOG.md');

    // Validate changelog for baz package
    expect(readChangelogMd(monoRepo.pathTo('packages/baz'))).toMatchSnapshot('baz CHANGELOG.md');

    // Validate grouped changelog for foo master package
    const groupedChangelogText = readChangelogMd(monoRepo.rootPath);
    expect(groupedChangelogText).toContain('- comment 1');
    expect(groupedChangelogText).not.toContain('- Bump baz');
    expect(groupedChangelogText).toMatchSnapshot('grouped CHANGELOG.md');
  });

  it('generates grouped changelog without dependent change entries where packages have normal changes and dependency changes', async () => {
    const monoRepo = monoRepoFactory.cloneRepository();
    monoRepo.commitChange('baz');
    generateChangeFiles([getChange('baz', 'comment 1')], monoRepo.rootPath);
    generateChangeFiles([getChange('bar', 'comment 1')], monoRepo.rootPath);

    const beachballOptions: Partial<BeachballOptions> = {
      path: monoRepo.rootPath,
      changelog: {
        groups: [
          {
            masterPackageName: 'foo',
            changelogPath: monoRepo.rootPath,
            include: ['packages/foo', 'packages/bar', 'packages/baz'],
          },
        ],
      },
    };

    const packageInfos = getPackageInfos(monoRepo.rootPath);
    const changes = readChangeFiles(beachballOptions as BeachballOptions, packageInfos);

    await writeChangelog(
      beachballOptions as BeachballOptions,
      changes,
      { bar: 'patch', baz: 'patch' },
      { bar: new Set(['baz']) },
      packageInfos
    );

    // Validate changelog for bar and baz packages
    expect(readChangelogMd(monoRepo.pathTo('packages/bar'))).toMatchSnapshot('bar CHANGELOG.md');
    expect(readChangelogMd(monoRepo.pathTo('packages/baz'))).toMatchSnapshot('baz CHANGELOG.md');

    // Validate grouped changelog for foo master package
    expect(readChangelogMd(monoRepo.rootPath)).toMatchSnapshot('grouped CHANGELOG.md');
  });

  it('generates correct grouped changelog when grouped change log is saved to the same dir as a regular changelog', async () => {
    const monoRepo = monoRepoFactory.cloneRepository();
    monoRepo.commitChange('foo');
    generateChangeFiles([getChange('foo', 'comment 1')], monoRepo.rootPath);

    monoRepo.commitChange('bar');
    generateChangeFiles([getChange('bar', 'comment 2')], monoRepo.rootPath);

    const beachballOptions: Partial<BeachballOptions> = {
      path: monoRepo.rootPath,
      changelog: {
        groups: [
          {
            masterPackageName: 'foo',
            changelogPath: monoRepo.pathTo('packages/foo'),
            include: ['packages/foo', 'packages/bar'],
          },
        ],
      },
    };

    const packageInfos = getPackageInfos(monoRepo.rootPath);
    const changes = readChangeFiles(beachballOptions as BeachballOptions, packageInfos);

    await writeChangelog(beachballOptions as BeachballOptions, changes, {}, {}, packageInfos);

    // Validate changelog for bar package
    expect(readChangelogMd(monoRepo.pathTo('packages/bar'))).toMatchSnapshot();

    // Validate grouped changelog for foo and bar packages
    expect(readChangelogMd(monoRepo.pathTo('packages/foo'))).toMatchSnapshot();
  });

  it('runs transform.changeFiles functions if provided', async () => {
    const editedComment: string = 'Edited comment for testing';
    const monoRepo = monoRepoFactory.cloneRepository();
    monoRepo.commitChange('foo');
    generateChangeFiles([getChange('foo', 'comment 1')], monoRepo.rootPath);

    monoRepo.commitChange('bar');
    generateChangeFiles([getChange('bar', 'comment 2')], monoRepo.rootPath);

    const beachballOptions: Partial<BeachballOptions> = {
      path: monoRepo.rootPath,
      command: 'change',
      transform: {
        changeFiles: (changeFile, changeFilePath, { command }) => {
          // For test, we will be changing the comment based on the package name
          if ((changeFile as ChangeInfo).packageName === 'foo') {
            (changeFile as ChangeInfo).comment = editedComment;
            (changeFile as ChangeInfo).command = command;
          }
          return changeFile as ChangeInfo;
        },
      },
      changelog: {
        groups: [
          {
            masterPackageName: 'foo',
            changelogPath: monoRepo.pathTo('packages/foo'),
            include: ['packages/foo', 'packages/bar'],
          },
        ],
      },
    };

    const packageInfos = getPackageInfos(monoRepo.rootPath);
    const changes = readChangeFiles(beachballOptions as BeachballOptions, packageInfos);

    // Verify that the comment of only the intended change file is changed
    for (const { change, changeFile } of changes) {
      if (changeFile.startsWith('foo')) {
        expect(change.comment).toBe(editedComment);
        expect(change.command).toEqual('change');
      } else {
        expect(change.comment).toBe('comment 2');
      }
    }
  });
});
