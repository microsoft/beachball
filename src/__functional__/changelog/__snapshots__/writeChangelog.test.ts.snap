// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`writeChangelog generates correct changelog in monorepo with groupChanges (grouped change FILES): bar CHANGELOG.json 1`] = `
Object {
  "entries": Array [
    Object {
      "comments": Object {
        "patch": Array [
          Object {
            "author": "test@testtestme.com",
            "comment": "comment from bar change ",
            "commit": "(sha1-0)",
            "package": "bar",
          },
        ],
      },
      "date": "(date)",
      "tag": "bar_v1.3.4",
      "version": "1.3.4",
    },
  ],
  "name": "bar",
}
`;

exports[`writeChangelog generates correct changelog in monorepo with groupChanges (grouped change FILES): bar CHANGELOG.md 1`] = `
"# Change Log - bar

This log was last generated on (date) and should not be manually modified.

<!-- Start content -->

## 1.3.4

(date)

### Patches

- comment from bar change  (test@testtestme.com)
"
`;

exports[`writeChangelog generates correct changelog in monorepo with groupChanges (grouped change FILES): foo CHANGELOG.json 1`] = `
Object {
  "entries": Array [
    Object {
      "comments": Object {
        "patch": Array [
          Object {
            "author": "test@testtestme.com",
            "comment": "comment 2",
            "commit": "(sha1-0)",
            "package": "foo",
          },
          Object {
            "author": "test@testtestme.com",
            "comment": "comment 1",
            "commit": "(sha1-1)",
            "package": "foo",
          },
          Object {
            "author": "test@testtestme.com",
            "comment": "additional comment 1",
            "commit": "(sha1-2)",
            "package": "foo",
          },
          Object {
            "author": "test@testtestme.com",
            "comment": "additional comment 2",
            "commit": "(sha1-3)",
            "package": "foo",
          },
        ],
      },
      "date": "(date)",
      "tag": "foo_v1.0.0",
      "version": "1.0.0",
    },
  ],
  "name": "foo",
}
`;

exports[`writeChangelog generates correct changelog in monorepo with groupChanges (grouped change FILES): foo CHANGELOG.md 1`] = `
"# Change Log - foo

This log was last generated on (date) and should not be manually modified.

<!-- Start content -->

## 1.0.0

(date)

### Patches

- comment 2 (test@testtestme.com)
- comment 1 (test@testtestme.com)
- additional comment 1 (test@testtestme.com)
- additional comment 2 (test@testtestme.com)
"
`;

exports[`writeChangelog generates correct changelog: changelog json 1`] = `
Object {
  "entries": Array [
    Object {
      "comments": Object {
        "patch": Array [
          Object {
            "author": "test@testtestme.com",
            "comment": "comment 2",
            "commit": "(sha1-0)",
            "package": "foo",
          },
          Object {
            "author": "test@testtestme.com",
            "comment": "comment 1",
            "commit": "(sha1-1)",
            "package": "foo",
          },
          Object {
            "author": "test@testtestme.com",
            "comment": "additional comment 1",
            "commit": "(sha1-2)",
            "package": "foo",
          },
          Object {
            "author": "test@testtestme.com",
            "comment": "additional comment 2",
            "commit": "(sha1-3)",
            "package": "foo",
          },
        ],
      },
      "date": "(date)",
      "tag": "foo_v1.0.0",
      "version": "1.0.0",
    },
  ],
  "name": "foo",
}
`;

exports[`writeChangelog generates correct changelog: changelog md 1`] = `
"# Change Log - foo

This log was last generated on (date) and should not be manually modified.

<!-- Start content -->

## 1.0.0

(date)

### Patches

- comment 2 (test@testtestme.com)
- comment 1 (test@testtestme.com)
- additional comment 1 (test@testtestme.com)
- additional comment 2 (test@testtestme.com)
"
`;

exports[`writeChangelog generates correct grouped changelog when grouped change log is saved to the same dir as a regular changelog 1`] = `
"# Change Log - bar

This log was last generated on (date) and should not be manually modified.

<!-- Start content -->

## 1.3.4

(date)

### Patches

- comment 2 (test@testtestme.com)
"
`;

exports[`writeChangelog generates correct grouped changelog when grouped change log is saved to the same dir as a regular changelog 2`] = `
"# Change Log - foo

This log was last generated on (date) and should not be manually modified.

<!-- Start content -->

## 1.0.0

(date)

### Patches

- \`bar\`
  - comment 2 (test@testtestme.com)
- \`foo\`
  - comment 1 (test@testtestme.com)
"
`;

exports[`writeChangelog generates correct grouped changelog: bar CHANGELOG.md 1`] = `
"# Change Log - bar

This log was last generated on (date) and should not be manually modified.

<!-- Start content -->

## 1.3.4

(date)

### Patches

- comment 3 (test@testtestme.com)
- comment 2 (test@testtestme.com)
"
`;

exports[`writeChangelog generates correct grouped changelog: foo CHANGELOG.md 1`] = `
"# Change Log - foo

This log was last generated on (date) and should not be manually modified.

<!-- Start content -->

## 1.0.0

(date)

### Patches

- comment 1 (test@testtestme.com)
"
`;

exports[`writeChangelog generates correct grouped changelog: grouped CHANGELOG.md 1`] = `
"# Change Log - foo

This log was last generated on (date) and should not be manually modified.

<!-- Start content -->

## 1.0.0

(date)

### Patches

- \`bar\`
  - comment 3 (test@testtestme.com)
  - comment 2 (test@testtestme.com)
- \`foo\`
  - comment 1 (test@testtestme.com)
"
`;

exports[`writeChangelog generates grouped changelog without dependent change entries where packages have normal changes and dependency changes: bar CHANGELOG.md 1`] = `
"# Change Log - bar

This log was last generated on (date) and should not be manually modified.

<!-- Start content -->

## 1.3.4

(date)

### Patches

- comment 1 (test@testtestme.com)
- Bump baz to v1.3.4
"
`;

exports[`writeChangelog generates grouped changelog without dependent change entries where packages have normal changes and dependency changes: baz CHANGELOG.md 1`] = `
"# Change Log - baz

This log was last generated on (date) and should not be manually modified.

<!-- Start content -->

## 1.3.4

(date)

### Patches

- comment 1 (test@testtestme.com)
"
`;

exports[`writeChangelog generates grouped changelog without dependent change entries where packages have normal changes and dependency changes: grouped CHANGELOG.md 1`] = `
"# Change Log - foo

This log was last generated on (date) and should not be manually modified.

<!-- Start content -->

## 1.0.0

(date)

### Patches

- \`bar\`
  - comment 1 (test@testtestme.com)
- \`baz\`
  - comment 1 (test@testtestme.com)
"
`;

exports[`writeChangelog generates grouped changelog without dependent change entries: bar CHANGELOG.md 1`] = `
"# Change Log - bar

This log was last generated on (date) and should not be manually modified.

<!-- Start content -->

## 1.3.4

(date)

### Patches

- Bump baz to v1.3.4
"
`;

exports[`writeChangelog generates grouped changelog without dependent change entries: baz CHANGELOG.md 1`] = `
"# Change Log - baz

This log was last generated on (date) and should not be manually modified.

<!-- Start content -->

## 1.3.4

(date)

### Patches

- comment 1 (test@testtestme.com)
"
`;

exports[`writeChangelog generates grouped changelog without dependent change entries: grouped CHANGELOG.md 1`] = `
"# Change Log - foo

This log was last generated on (date) and should not be manually modified.

<!-- Start content -->

## 1.0.0

(date)

### Patches

- \`baz\`
  - comment 1 (test@testtestme.com)
"
`;
